File: Categories/Dynamic_Programming/329_longestIncreasingPathInAMatrix.cs
public class Solution {
    int num_rows, num_cols;
    int[][] dp;
    int[][] directions;

    private int dfs(ref int[][] matrix, int row, int col) {
        if (dp[row][col] != 0) {
            return dp[row][col];
        }

        dp[row][col] = 1;
        int nrow, ncol, drow, dcol;
        foreach(int[] direction in directions) {
            drow = direction[0];
            dcol = direction[1];
            nrow = row + drow;
            ncol = col + dcol;

            if (
                nrow < 0 ||
                nrow >= num_rows ||
                ncol < 0 ||
                ncol >= num_cols
            ) {
                continue;
            }

            if ( matrix[nrow][ncol] >= matrix[row][col] ) {
                continue;
            }
            
            dp[row][col] = Math.Max(
                dp[row][col],
                dfs(ref matrix, nrow, ncol) + 1
            );
        }

        return dp[row][col];
    }

    public int LongestIncreasingPath(int[][] matrix) {
        num_rows = matrix.Length;
        num_cols = matrix[0].Length;

        // dp[row][col]: longest increasing path ending at position (row, col)
        dp = new int[num_rows][];
        for (int row = 0; row < num_rows; row++) {
            dp[row] = new int[num_cols];
        }

        directions = new int [][] 
        {
            new int[] {+1, 0},
            new int[] {0, +1},
            new int[] {-1, 0},
            new int[] {0, -1}
        };

        int ret = 0;
        for (int row = 0; row < num_rows; row++) {
            for (int col = 0; col < num_cols; col++) {
                ret = Math.Max(
                    ret,
                    dfs(ref matrix, row, col)
                );
            }
        }

        return ret;
    }
}

File: Categories/Dynamic_Programming/329_longestIncreasingPathInAMatrix.py
from typing import *
from collections import defaultdict

class Solution:
    def dfs(self, row: int, col: int) -> int:
        if self.dp[row][col] != 0:
            return self.dp[row][col]

        self.dp[row][col] = 1
        for direction in self.directions:
            drow, dcol = direction
            nrow, ncol = row + drow, col + dcol
            if (
                nrow < 0 or nrow >= self.num_rows or
                ncol < 0 or ncol >= self.num_cols
            ):
                continue
            if self.matrix[nrow][ncol] >= self.matrix[row][col]:
                continue
            
            self.dp[row][col] = max(
                self.dp[row][col],
                self.dfs(nrow, ncol) + 1,
            )
        
        return self.dp[row][col]
        
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        # dp[row][col] stricktly increasing path ending at position (row, col)
        self.matrix = matrix
        self.num_rows, self.num_cols = len(matrix), len(matrix[0])
        self.dp = defaultdict(lambda : defaultdict(int))
        self.directions = [
            [+1, 0],
            [0, +1],
            [-1, 0],
            [0, -1],
        ]

        ret = 0
        for row in range(self.num_rows):
            for col in range(self.num_cols):
                ret = max(
                    ret,
                    self.dfs(row, col)
                )
        
        return ret
                